adapters/beachfront/beachfront.go

```diff
@@ -111,300 +96,388 @@ type BeachfrontResponseSlot struct {
+		} else {
+			slot.Bidfloor = beachfrontExt.BidFloor
+		}
+
+		slot.Slot = request.Imp[i].ID
+
+		for j := 0; j < len(request.Imp[i].Banner.Format); j++ {
 
-			beachfrontReq.Slots[bannerImpsIndex].Bidfloor = beachfrontExt.BidFloor
-			beachfrontReq.Slots[bannerImpsIndex].Slot = req.Imp[bannerImpsIndex].ID
-			beachfrontReq.Slots[bannerImpsIndex].Id = beachfrontExt.AppId
+			slot.Sizes = append(slot.Sizes, beachfrontSize{
+				H: request.Imp[i].Banner.Format[j].H,
+				W: request.Imp[i].Banner.Format[j].W,
+			})
 		}
 
-		imps++
+		bfr.Slots = append(bfr.Slots, slot)
 	}
 
-	// Just take the last one... I guess?
-	if req.Imp[bannerImpsIndex].Secure != nil {
-		beachfrontReq.Secure = *req.Imp[bannerImpsIndex].Secure
+	if len(bfr.Slots) == 0 {
+		return bfr, errs
 	}
 
-	if req.User != nil {
-		beachfrontReq.User.ID = req.User.ID
-		beachfrontReq.User.BuyerUID = req.User.BuyerUID
+	if request.Device != nil {
+		bfr.IP = getIP(request.Device.IP)
+		bfr.DeviceModel = request.Device.Model
+		bfr.DeviceOs = request.Device.OS
+		if request.Device.DNT != nil {
+			bfr.Dnt = *request.Device.DNT
+		}
+		if request.Device.UA != "" {
+			bfr.UA = request.Device.UA
+		}
 	}
 
-	if req.App != nil {
-		beachfrontReq.Domain = req.App.Domain
-		beachfrontReq.Page = req.App.ID
-	} else {
-		protoUrl := strings.Split(req.Site.Page, "//")
-		var domainPage string
-		// Resolves a panic for any Site.Page that does not include the protocol
-		if len(protoUrl) > 1 {
-			domainPage = protoUrl[1]
+	if request.App != nil {
+		bfr.Page = request.App.Bundle
+		if request.App.Domain == "" {
+			bfr.Domain = getDomain(request.App.Domain)
 		} else {
-			domainPage = protoUrl[0]
+			bfr.Domain = request.App.Domain
+		}
+
+		bfr.IsMobile = 1
+	} else if request.Site.Page != "" {
+		bfr.Page = request.Site.Page
+		if request.Site.Domain == "" {
+			bfr.Domain = getDomain(request.Site.Page)
+		} else {
+			bfr.Domain = request.Site.Domain
+		}
+
+		bfr.IsMobile = 0
+	}
+
+	bfr.Secure = isSecure(bfr.Page)
+
+	if request.User != nil && request.User.ID != "" {
+		if bfr.User.ID == "" {
+			bfr.User.ID = request.User.ID
 		}
-		beachfrontReq.Domain = strings.Split(domainPage, "/")[0]
-		beachfrontReq.Page = req.Site.Page
 	}
 
-	beachfrontReq.RequestID = req.ID
+	if request.User != nil && request.User.BuyerUID != "" {
+		if bfr.User.BuyerUID == "" {
+			bfr.User.BuyerUID = request.User.BuyerUID
+		}
+	}
 
-	return beachfrontReq, errs, imps
+	bfr.RequestID = request.ID
+	bfr.AdapterName = beachfrontAdapterName
+	bfr.AdapterVersion = beachfrontAdapterVersion
+
+	if request.Imp[0].Secure != nil {
+		bfr.Secure = *request.Imp[0].Secure
+	}
+
+	return bfr, errs
 }
 
-func getVideoRequest(req *openrtb.BidRequest) (BeachfrontVideoRequest, []error, int) {
-	var videoImpsIndex = 0
-	var beachfrontReq = NewBeachfrontVideoRequest()
-	var errs = make([]error, 0, len(req.Imp))
-	var imps = 0
+func guessDeviceType(request *openrtb.BidRequest) openrtb.DeviceType {
+	if request.Site != nil {
```

Comment:
Would it be better to base the guess on the device screen size? This really just breaks down to app or site.

Reply:
That would be a good test, as well as UA (see 
a6d761d). I think a full implementation would try by UA, then screen size, and App/Site as a fallback, but now I'm being asked to just do this much and just consider types 1 and 2. I think it's a matter of beachfront wanting to audit how they are implementing deviceTypes on their back end. So semantically it would make sense for me to rename this function as it is not a "guess".